resource/api_model_test.go

```diff
@@ -0,0 +1,549 @@
+	test.That(t, resConfig.API.Type.Namespace, test.ShouldEqual, "not_rdk")
+	test.That(t, resConfig.API.Type.Name, test.ShouldEqual, "component")
+}
+
+func TestServiceAdjustPartialNames(t *testing.T) {
+	resConfig := Config{
+		API: API{
+			SubtypeName: "nav",
+		},
+	}
+
+	// Running `AdjustPartialNames` on a service should fill in all of the remaining `API` and
+	// `Model` fields.
+	resConfig.AdjustPartialNames("service")
+	test.That(t, resConfig.Name, test.ShouldEqual, "builtin")
+	test.That(t, resConfig.API.Type.Namespace, test.ShouldEqual, "rdk")
+	test.That(t, resConfig.API.Type.Name, test.ShouldEqual, "service")
+	test.That(t, resConfig.API.SubtypeName, test.ShouldEqual, "nav")
+	test.That(t, resConfig.Model.Family.Namespace, test.ShouldEqual, "rdk")
+	test.That(t, resConfig.Model.Family.Name, test.ShouldEqual, "builtin")
+	test.That(t, resConfig.Model.Name, test.ShouldEqual, "builtin")
+}
+
+func TestAPIStringParsing(t *testing.T) {
+	// Test that a colon-delimited triplet string for the `API` is parsed into its
+	// namespace/type/subtype tuple.
+	var resConfig Config
+	err := resConfig.UnmarshalJSON([]byte(`
+{
+  "name": "legacy",
+  "type": "board",
+  "model": "fake",
+  "api": "namespace:component_type:subtype"
+}`))
+	test.That(t, err, test.ShouldBeNil)
+
+	test.That(t, resConfig.API.Type.Namespace, test.ShouldEqual, "namespace")
+	test.That(t, resConfig.API.Type.Name, test.ShouldEqual, "component_type")
+	test.That(t, resConfig.API.SubtypeName, test.ShouldEqual, "subtype")
+
+	// Test that a malformed string gives a string specific error message.
+	err = resConfig.UnmarshalJSON([]byte(`
+{
+  "name": "legacy",
+  "type": "board",
+  "model": "fake",
+  "api": "double::colons::are_bad"
+}`))
+	test.That(t, err, test.ShouldNotBeNil)
+	test.That(t, err.Error(), test.ShouldContainSubstring, "not a valid API config string.")
+}
+
+func TestAPIObjectParsing(t *testing.T) {
+	// Test that an API object with all three types parses correctly.
+	var resConfig Config
+	err := resConfig.UnmarshalJSON([]byte(`
+{
+  "name": "legacy",
+  "type": "board",
+  "model": "fake",
+  "api": {
+    "namespace": "namespace",
+    "type": "component",
+    "subtype": "subtype"
+  }
+}`))
+	test.That(t, err, test.ShouldBeNil)
+
+	test.That(t, resConfig.API.Type.Namespace, test.ShouldEqual, "namespace")
+	test.That(t, resConfig.API.Type.Name, test.ShouldEqual, "component")
+	test.That(t, resConfig.API.SubtypeName, test.ShouldEqual, "subtype")
+
+	// Test that a malformed API object does not parse. Without enumerating all of the ways an API
+	// object can be malformed.
+	err = resConfig.UnmarshalJSON([]byte(`
+{
+  "name": "legacy",
+  "type": "board",
+  "model": "fake",
+  "api": {}
+}`))
+	test.That(t, err, test.ShouldNotBeNil)
+	test.That(t, err.Error(), test.ShouldContainSubstring, "namespace field for resource missing or invalid")
+}
+
+func TestModel(t *testing.T) {
+	for _, tc := range []struct {
+		TestName  string
```

Comment:
I hope you don't mind me not feeling inclined to change this :)

But your comment did pique my curiosity: can anonymous structs even be exported in a way such that another file (much less package) can access its members?

Reply:
Yeah no worries leaving as-is.

> can anonymous structs even be exported in a way such that another file (much less package) can access its members?

Not that I know of since they're not named outside of their inline declaration. So no real danger of anything importing this API, just a code smell.